python/cudf/cudf/core/multiindex.py

```diff
@@ -1051,7 +994,10 @@ def __getitem__(self, index):
             return match
 
     def to_frame(self, index=True, name=None):
-        df = self._source_data
+        # TODO: Currently this function makes a shallow copy, which is
+        # incorrect. We want to make a deep copy, otherwise further
+        # modifications of the resulting DataFrame will affect the MultiIndex.
+        df = cudf.DataFrame._from_data(data=self._data)
```

Comment:
What's preventing you from making a deep copy of `self._data` here?

Reply:
The fact that I'm currently using `to_frame` internally in some places as a replacement for `_source_data`, so making this always copy would slow a lot of things down. I could patch in an internal API that accepts a `copy` argument, but I figured that it would be OK to let this issue persist for a little bit while I figure out which cases should be copies and which shouldn't. IIRC I found that some APIs that _should_ be returning deep copies are currently returning views. I just decided to limit the scope of this PR and push that to a downstream PR.

Reply:
:+1: